src/outside/codegen.cc

```diff
@@ -0,0 +1,229 @@
+
+}}} */
+
+#include "codegen.hh"
+
+
+std::list<Symbol::NT*> *NTs_to_report(const AST &ast) {
+  /* define which non-terminals shell be reported to the user
+   * order of user arguments (--outside_grammar) shall take precedence over
+   * NTs as occurring in source code of grammar.
+   * - User shell be warned, if outside version of NT has not been generated.
+   *   This happens for NTs without rhs NTs.
+   * - User was already informed about NTs he/she requested but are not part of
+   *   the grammar. */
+  std::list<Symbol::NT*> *report_nts = new std::list<Symbol::NT*>();
+
+  // iterate through user arguments
+  for (std::vector<std::string>::const_iterator name_param =
+       ast.get_outside_nt_list()->begin();
+      name_param != ast.get_outside_nt_list()->end(); ++name_param) {
+    if (name_param->compare(OUTSIDE_ALL) == 0) {
+      // edge case: user requested to report ALL non terminals, by providing the
+      // keyword "ALL"
+      for (std::list<Symbol::NT*>::const_iterator i =
+           ast.grammar()->nts().begin(); i != ast.grammar()->nts().end(); ++i) {
+        // skip outside NTs
+        if ((*i)->is_partof_outside()) {
+          continue;
+        }
+        bool already_in = false;
+        for (std::list<Symbol::NT*>::iterator rnt = report_nts->begin();
+             rnt != report_nts->end(); ++rnt) {
+          if (*rnt == *i) {
+            already_in = true;
+            break;
+          }
+        }
+        if (already_in == false) {
+          report_nts->push_back(*i);
+        }
+      }
+    } else {
+      report_nts->push_back(dynamic_cast<Symbol::NT*>(
+          ast.grammar()->NTs.find(*name_param)->second));
+    }
+  }
+
+  std::list<Symbol::NT*> nts_no_outside;
+  for (std::list<Symbol::NT*>::iterator i = report_nts->begin();
+       i != report_nts->end(); ++i) {
+    hashtable<std::string, Symbol::Base*>::iterator outside =
+        ast.grammar()->NTs.find(OUTSIDE_NT_PREFIX + *(*i)->name);
+    if (outside == ast.grammar()->NTs.end()) {
+      nts_no_outside.push_back(*i);
+    }
+  }
+  if (nts_no_outside.size() > 0) {
+    std::string msg = "Outside generation of your grammar will NOT lead to out"
+                      "side versions of the following " +
+                      std::to_string(nts_no_outside.size()) +
+                      " non-terminals: ";
+    for (std::list<Symbol::NT*>::iterator i = nts_no_outside.begin();
+         i != nts_no_outside.end(); ++i) {
+      msg += *(*i)->name;
+      if (i != std::next(nts_no_outside.end())) {
+        msg += ", ";
+      }
+    }
+    msg += ", as they have no non-terminal on their right hand sides.";
+    Log::instance()->warning(nts_no_outside.front()->location, msg);
+  }
+
+  return report_nts;
+}
+
+
+void print_insideoutside_report_fn(Printer::Cpp &stream, const AST &ast) {
+  bool old_in_class = stream.in_class;
+  stream.in_class = true;
+
+  std::list<Symbol::NT*> *report_nts = NTs_to_report(ast);
```

Comment:
`report_nts` should be deleted at the end of the `print_insideoutside_report_fn` function.

Reply:
Actually, if I'm not mistaken, the code for the `report_insideoutside` function is directly added to the output file(stream) at the end of `print_insideoutside_report_fn`, right? In that case, literally every heap-allocated object can be deleted at the end of the function. I definitely can't point all of those out individually. If you want I can try to adjust this function myself to properly delete everything at the end. Shouldn't take very long, its just too much work to point every individual example out.

Reply:
your assumption is correct, `print_insideoutside_report_fn` produces the code that reports the DP cell values after inside and outside pass. Your contribution is highly appreciated!